pycbc/inference/sampler/multinest.py

```diff
@@ -0,0 +1,427 @@
+        while not done:
+            logging.info("Running sampler for {} to {} iterations".format(
+                self.niterations, self.niterations + iterinterval))
+            # run multinest
+            self.solve(self.ns_loglikelihood, self.ns_prior, self._ndim,
+                       n_live_points=self.nlivepoints,
+                       evidence_tolerance=self._ztol,
+                       sampling_efficiency=self._eff,
+                       importance_nested_sampling=self._ins,
+                       max_iter=iterinterval,
+                       outputfiles_basename=outputfiles_basename,
+                       multimodal=False, verbose=True)
+            # parse results from multinest output files
+            nest_stats = a.get_mode_stats()
+            self._logz = nest_stats['nested sampling global log-evidence']
+            self._dlogz = nest_stats['nested sampling global log-evidence error']
+            if self._ins:
+                self._importance_logz = nest_stats['nested importance sampling global log-evidence']
+                self._importance_dlogz = nest_stats['nested importance sampling global log-evidence error']
+            self._samples = self.get_posterior_samples()[:,:-1]
+            logging.info("Have {} posterior samples".format(self._samples.shape[0]))
+            # update the itercounter
+            self._itercount = self._itercount + iterinterval
+            # make sure there's at least 1 posterior sample
+            if self._samples.shape[0] == 0:
+                continue
+            # dump the current results
+            self.checkpoint()
+            # check if we're finished
+            done = self.check_if_finished()
+
+    def solve(self, LogLikelihood, Prior, n_dims, **kwargs):
+	kwargs['n_dims'] = n_dims
+	files_temporary = False
+	if 'outputfiles_basename' not in kwargs:
+		files_temporary = True
+		tempdir = tempfile.mkdtemp('pymultinest')
+		kwargs['outputfiles_basename'] = tempdir + '/'
+	outputfiles_basename = kwargs['outputfiles_basename']
+	def SafePrior(cube, ndim, nparams):
+	    try:
+	        a = numpy.array([cube[i] for i in range(n_dims)])
+		b = Prior(a)
+		for i in range(n_dims):
+		    cube[i] = b[i]
+	    except Exception as e:
+		import sys
+		sys.stderr.write('ERROR in prior: %s\n' % e)
+	        sys.exit(1)
+
+	def SafeLoglikelihood(cube, ndim, nparams, lnew):
+	    try:
+	        a = numpy.array([cube[i] for i in range(n_dims)])
+	        l = float(LogLikelihood(a))
+		if not numpy.isfinite(l):
+		    import sys
+		    sys.stderr.write('WARNING: loglikelihood not finite: %f\n' % (l))
+		    sys.stderr.write('         for parameters: %s\n' % a)
+		    sys.stderr.write('         returned very low value instead\n')
+		    return -1e100
+		return l
+	    except Exception as e:
+		import sys
+		sys.stderr.write('ERROR in loglikelihood: %s\n' % e)
+		sys.exit(1)
+
+	kwargs['LogLikelihood'] = SafeLoglikelihood
+	kwargs['Prior'] = SafePrior
+	pymultinest.run(**kwargs)
```

Comment:
@dfinstad On second thought, I'm fine to keep the `ns_likelihood` and prior functions as class methods of the sampler, rather than defining a new class. So ignore my last statement. I would prefer that functions not be defined in functions though, so move those definitions outside of the functions.

Do fix the logposterior -> loglikelihood issue though.